### PR TITLE
Register feature now works with the rails server

### PR DESCRIPTION
**Pull request by Willhs**
Quite a big merge.
- Changes to user service, register controller and their tests.
- Added a bit of grunt and jasmine config so that server address will change for build/test type dev or prod

### CODE REVIEW

File: test/spec/services/user.js [hunk 1]
```diff
@@ -5,78 +5,122 @@ describe('Service: user', function () {
...
-    user.attemptRegister('Alice');
-    var promise = user.attemptRegister('John');
-    var spy = jasmine.createSpy('different users spy');
-    promise.then(angular.noop, spy);
+    var name1 = "will";
+    var name2 = "monkey";
+    spyOn(http, "get").and.returnValue(fakeGetPromise(true));
+    spyOn(http, "post").and.returnValue(fakePostPromise(true));
+    var successSpy = jasmine.createSpy('succeed spy');
+    var failSpy = jasmine.createSpy('fail spy');
```
Comment (jman48) [inline test/spec/services/user.js]: There does not seem to be any use for this spy. Just remove it if you are not using it.

File: app/scripts/controllers/register.js [hunk 2]
```diff
@@ -9,13 +9,23 @@
...
-            $scope.errorMessage = errorMsg;
-        });
-      };
+          function (errorMsg) {
+            $scope.errorMessage = "Request error: " + errorMsg;
+          });
+      }
+    };
+
+    $scope.test = function () {
```
Comment (jman48) [inline app/scripts/controllers/register.js]: Don't see why this exists. If it was for some debugging don't forget to remove it before making a merge request.

File: app/scripts/services/user.js [hunk 3]
```diff
@@ -31,39 +34,65 @@ angular.module('bballApp')
...
+      if (!username) {
         defer.reject("Username is not valid");
-      if (users.hasOwnProperty(username.toLowerCase())){
+      } else if (users.hasOwnProperty(username.toLowerCase())) {
         currentUser = users[username.toLowerCase()];
-        $rootScope.userName = username;
         defer.resolve(username);
-      }
-      else defer.reject("Username is not registered");
+        $rootScope.userName = username;
```
Comment (jman48) [inline app/scripts/services/user.js]: Should avoid defining anything on the rootScope. Use a service to store the username.

File: test/spec/services/user.js [hunk 4]
```diff
@@ -5,78 +5,122 @@ describe('Service: user', function () {
...
+    expect(successSpy).toHaveBeenCalledWith(name1);
+    expect(successSpy).toHaveBeenCalledWith(name2);
   });
 
-<<<<<<< HEAD
-  it('should fail when username field is blank', function() {
-    var promise = user.attemptRegister();
-    var spy = jasmine.createSpy('blank username');
-    promise.then(angular.noop, spy);
+  // todo: uncomment when post is implemented
```
Comment (jman48) [inline test/spec/services/user.js]: Should uncomment it or remove it.

**Comment by jman48** [comment]
Looking good. Just a few things that I commented on.

File: app/scripts/services/user.js (hunk 3 above)
Comment (Willhs) [inline app/scripts/services/user.js]: This is change to the login feature. I'm not sure how it got into this branch. Are you OK with leaving this in until the next pull request from login_feature?

File: app/scripts/controllers/register.js (hunk 2 above)
Comment (Willhs) [inline app/scripts/controllers/register.js]: Now resolved.

File: app/scripts/services/user.js (hunk 3 above)
Comment (thomsedavi) [inline app/scripts/services/user.js]: I had some difficulty difficulty getting the html to retrieve information from the user service so it was suggested that I put the information on rootScope instead, I have been trying to fix this but may need some assistance.

File: test/spec/services/user.js (hunk 4 above)
Comment (Willhs) [inline test/spec/services/user.js]: Resolved.

**Pull request closed without merging**